[Quant][fx] Lower reference conv[1-3]d module

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69228




Implement lowering logic for reference conv modules,
similar to https://github.com/pytorch/pytorch/pull/65723.

Differential Revision: [D32890743](https://our.internmc.facebook.com/intern/diff/D32890743/)

Differential Revision: [D32890743](https://our.internmc.facebook.com/intern/diff/D32890743)

andrewor14 committed:
Temp

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -35,31 +48,27 @@ def _lower_ref_linear_module(model: QuantizedGraphModule) -> QuantizedGraphModul
 
         # this can be removed if we add support for constants in is_match
-        if dtype != torch.quint8:
+        if q_class == torch.nn.quantized.Linear and dtype != torch.quint8:
```
this check is valid for both linear and conv I think

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -3199,6 +3199,38 @@ def forward(self, x):
         self.assertEqual(out_ref, out)
 
+    def test_conv_lowering(self):
+        convs = [nn.Conv1d, nn.Conv2d, nn.Conv3d]
```
nit: we typically use a map from dim to conv class, e.g.: https://github.com/pytorch/pytorch/blob/master/test/quantization/fx/test_quantize_fx.py#L3619, that might be clearer

andrewor14 committed:
Update on "[WIP][Quant][fx] Lower reference conv[1-3]d module"


Summary:

Test Plan:

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T105680980

[ghstack-poisoned]

andrewor14 committed:
Update on "[WIP][Quant][fx] Lower reference conv[1-3]d module"


Summary:

Test Plan:

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T105680980

[ghstack-poisoned]

andrewor14 committed:
Update on "[WIP][Quant][fx] Lower reference conv[1-3]d module"


Summary: Implement lowering logic for reference conv modules,
similar to https://github.com/pytorch/pytorch/pull/65723.

Test Plan: python test/test_quantization.py TestQuantizeFx.test_conv_lowering

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T105680980

Tags: quant, fx

[ghstack-poisoned]

andrewor14 committed:
Update on "[WIP][Quant][fx] Lower reference conv[1-3]d module"


Summary: Implement lowering logic for reference conv modules,
similar to https://github.com/pytorch/pytorch/pull/65723.

Test Plan: python test/test_quantization.py TestQuantizeFx.test_conv_lowering

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T105680980

Tags: quant, fx

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Lower reference conv[1-3]d module"


Summary: Implement lowering logic for reference conv modules,
similar to https://github.com/pytorch/pytorch/pull/65723.

Test Plan: python test/test_quantization.py TestQuantizeFx.test_conv_lowering

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T105680980

Tags: quant, fx

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -6,13 +6,26 @@
+    torch.nn.quantized._reference.Conv3d: torch.nn.quantized.Conv3d,
+}
+
+def _lower_ref(model: QuantizedGraphModule, ref_class: type) -> QuantizedGraphModule:
```
nit: maybe rename to `_lower_weighted_ref_module` to be more explicit

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -6,13 +6,26 @@
-        (torch.nn.quantized._reference.Linear, "dequantize"),
-        MatchAllNode, MatchAllNode, MatchAllNode)
+# Mapping from reference module class to the replacement quantized module class for lowering
+_lower_module_map = {
```
nit: we typically use something like `LOWER_MODULE_MAP` to indicate this is a constant map

jerryzh168 reviewed (approved):
Looks great!

jerryzh168 commented:
please import the PR to phabricator with ghimport as well

andrewor14 committed:
Update on "[Quant][fx] Lower reference conv[1-3]d module"


Summary: Implement lowering logic for reference conv modules,
similar to https://github.com/pytorch/pytorch/pull/65723.

Test Plan: python test/test_quantization.py TestQuantizeFx.test_conv_lowering

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T105680980

Tags: quant, fx

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32890743).

andrewor14 committed:
Update on "[Quant][fx] Lower reference conv[1-3]d module"


Summary: Implement lowering logic for reference conv modules,
similar to https://github.com/pytorch/pytorch/pull/65723.

Test Plan: python test/test_quantization.py TestQuantizeFx.test_conv_lowering

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T105680980

Tags: quant, fx

Differential Revision: [D32890743](https://our.internmc.facebook.com/intern/diff/D32890743)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32890743).

andrewor14 committed:
Update on "[Quant][fx] Lower reference conv[1-3]d module"


Summary: Implement lowering logic for reference conv modules,
similar to https://github.com/pytorch/pytorch/pull/65723.

Test Plan: python test/test_quantization.py TestQuantizeFx.test_conv_lowering

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T105680980

Tags: quant, fx

Differential Revision: [D32890743](https://our.internmc.facebook.com/intern/diff/D32890743)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32890743).

andrewor14 committed:
Update on "[Quant][fx] Lower reference conv[1-3]d module"


Summary: Implement lowering logic for reference conv modules,
similar to https://github.com/pytorch/pytorch/pull/65723.

Test Plan: python test/test_quantization.py TestQuantizeFx.test_conv_lowering

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T105680980

Tags: quant, fx

Differential Revision: [D32890743](https://our.internmc.facebook.com/intern/diff/D32890743)

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Lower reference conv[1-3]d module"





Implement lowering logic for reference conv modules,
similar to https://github.com/pytorch/pytorch/pull/65723.

Differential Revision: [D32890743](https://our.internmc.facebook.com/intern/diff/D32890743/)

Differential Revision: [D32890743](https://our.internmc.facebook.com/intern/diff/D32890743)

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Lower reference conv[1-3]d module"





Implement lowering logic for reference conv modules,
similar to https://github.com/pytorch/pytorch/pull/65723.

Differential Revision: [D32890743](https://our.internmc.facebook.com/intern/diff/D32890743/)

Differential Revision: [D32890743](https://our.internmc.facebook.com/intern/diff/D32890743)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32890743).

andrewor14 committed:
Update on "[Quant][fx] Lower reference conv[1-3]d module"





Implement lowering logic for reference conv modules,
similar to https://github.com/pytorch/pytorch/pull/65723.

Differential Revision: [D32890743](https://our.internmc.facebook.com/intern/diff/D32890743/)

Differential Revision: [D32890743](https://our.internmc.facebook.com/intern/diff/D32890743)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32890743).
